pilot/pkg/config/kube/gateway/deploymentcontroller.go

```diff
@@ -0,0 +1,309 @@
+
+	defer d.queue.Done(key)
+
+	err := d.Reconcile(key.(types.NamespacedName))
+	if err != nil {
+		if d.queue.NumRequeues(key) < 5 {
+			log.Errorf("error handling %v, retrying: %v", key, err)
+			d.queue.AddRateLimited(key)
+		} else {
+			log.Errorf("error handling %v, and retry budget exceeded: %v", key, err)
+		}
+	}
+	return true
+}
+
+// Reconcile takes in the name of a Gateway and ensures the cluster is in the desired state
+func (d *DeploymentController) Reconcile(req types.NamespacedName) error {
+	log := log.WithLabels("gateway", req)
+
+	gw, err := d.client.GatewayAPIInformer().Gateway().V1alpha2().Gateways().Lister().Gateways(req.Namespace).Get(req.Name)
+	if err != nil || gw == nil {
+		log.Errorf("unable to fetch Gateway: %v", err)
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return controllers.IgnoreNotFound(err)
+	}
+
+	switch gw.Spec.GatewayClassName {
+	case DefaultClassName:
+		return d.configureIstioGateway(log, *gw)
+	}
+
+	return nil
+}
+
+func (d *DeploymentController) configureIstioGateway(log *istiolog.Scope, gw gateway.Gateway) error {
+	// If user explicitly sets addresses, we are assuming they are pointing to an existing deployment.
+	// We will not manage it in this case
+	if !isManaged(&gw.Spec) {
+		log.Debug("skip unmanaged gateway")
+		return nil
+	}
+	log.Info("reconciling")
+
+	if err := d.ApplyTemplate("service.yaml", serviceInput{gw, extractServicePorts(gw)}); err != nil {
+		return fmt.Errorf("update service: %v", err)
+	}
+	log.Info("service updated")
+	if gw.Annotations[ExportService] == "true" {
+		if err := d.ApplyTemplate("service-export.yaml", gw); err != nil {
+			return fmt.Errorf("update service-export: %v", err)
+		}
+		log.Info("service-export updated")
+	}
+
+	if err := d.ApplyTemplate("deployment.yaml", gw); err != nil {
+		return fmt.Errorf("update deployment: %v", err)
+	}
+	log.Info("deployment updated")
+
+	gws := &gateway.Gateway{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       gvk.KubernetesGateway.Kind,
+			APIVersion: gvk.KubernetesGateway.Group + "/" + gvk.KubernetesGateway.Version,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      gw.Name,
+			Namespace: gw.Namespace,
+		},
+		Status: gateway.GatewayStatus{
+			Conditions: setConditions(gw.Generation, nil, map[string]*condition{
+				string(gateway.GatewayConditionScheduled): {
+					reason:  "ResourcesAvailable",
+					message: "Deployed gateway to the cluster",
+				},
+			}),
+		},
+	}
+	if err := d.ApplyObject(gws, "status"); err != nil {
+		return fmt.Errorf("update gateway status: %v", err)
+	}
+	log.Info("gateway updated")
+	return nil
+}
+
+// ApplyTemplate renders a template with the given input and (server-side) applies the results to the cluster.
+func (d *DeploymentController) ApplyTemplate(template string, input interface{}, subresources ...string) error {
```

Comment:
Seems like a pretty generic function. 
- don't we have it already ?
- should we move to /pkg ?

Reply:
We don't. The only thing really apply templates to cluster is the operator code, which use controller-runtime which is incompatible with our needs for Istiod (discussed at length in the DeploymentController comments).

There is the new `controllers` package so moving it there seems reasonable